File: example-servers/python/flask/src/services/cohere.py
import requests
import os

# Make sure to set the COHERE_API_KEY environment variable in a .env file (create if does not exist) - see .env.example

class Cohere:
    def chat(self, body):
        headers = {
            "Content-Type": "application/json",
            "Authorization": "Bearer " + os.getenv("COHERE_API_KEY")
        }
        chat_body = self.create_chat_body(body)
        response = requests.post(
            "https://api.cohere.ai/v1/chat", json=chat_body, headers=headers)
        json_response = response.json()
        if "message" in json_response:
            raise Exception(json_response["message"])
        # Sends response back to Deep Chat using the Response format:
        # https://deepchat.dev/docs/connect/#Response
        return {"text": json_response["text"]}

    @staticmethod
    def create_chat_body(body):
        # Text messages are stored inside request body using the Deep Chat JSON format:
        # https://deepchat.dev/docs/connect
        return {
            'query': body['messages'][len(body['messages']) - 1]['text'],
            'chat_history': [
                {
                    'user_name': 'CHATBOT' if message['role'] == 'ai' else 'USER',
                    'text': message['text']
                } for message in body['messages'][:-1]
            ],
        }

    def generate_text(self, body):
        headers = {
            "Content-Type": "application/json",
            "Authorization": "Bearer " + os.getenv("COHERE_API_KEY")
        }
        # Text messages are stored inside request body using the Deep Chat JSON format:
        # https://deepchat.dev/docs/connect
        generation_body = {"prompt": body["messages"][0]["text"]}
        response = requests.post(
            "https://api.cohere.ai/v1/generate", json=generation_body, headers=headers)
        json_response = response.json()
        if "message" in json_response:
            raise Exception(json_response["message"])
        result = json_response["generations"][0]["text"]
        # Sends response back to Deep Chat using the Response format:
        # https://deepchat.dev/docs/connect/#Response
        return {"text": result}
    
    def summarize_text(self, body):
        headers = {
            "Content-Type": "application/json",
            "Authorization": "Bearer " + os.getenv("COHERE_API_KEY")
        }
        # Text messages are stored inside request body using the Deep Chat JSON format:
        # https://deepchat.dev/docs/connect
        summarization_body = {"text": body["messages"][0]["text"]}
        response = requests.post(
            "https://api.cohere.ai/v1/summarize", json=summarization_body, headers=headers)
        json_response = response.json()
        if "message" in json_response:
            raise Exception(json_response["message"])
        result = json_response["summary"]
        # Sends response back to Deep Chat using the Response format:
        # https://deepchat.dev/docs/connect/#Response
        return {"text": result}

File: example-servers/node/express/src/services/cohere.ts
import {NextFunction, Request, Response} from 'express';
import https from 'https';

// Make sure to set the COHERE_API_KEY environment variable in a .env file (create if does not exist) - see .env.example

export class Cohere {
  public static async chat(body: Request['body'], res: Response, next: NextFunction) {
    // Text messages are stored inside request body using the Deep Chat JSON format:
    // https://deepchat.dev/docs/connect
    const chatBody = Cohere.createChatBody(body);
    const req = https.request(
      'https://api.cohere.ai/v1/chat',
      {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
          Authorization: 'Bearer ' + process.env.COHERE_API_KEY,
        },
      },
      (reqResp) => {
        let data = '';
        reqResp.on('error', next); // forwarded to error handler middleware in ErrorUtils.handle
        reqResp.on('data', (chunk) => {
          data += chunk;
        });
        reqResp.on('end', () => {
          const result = JSON.parse(data);
          if (result.message) {
            next(result); // forwarded to error handler middleware in ErrorUtils.handle
          } else {
            // Sends response back to Deep Chat using the Response format:
            // https://deepchat.dev/docs/connect/#Response
            res.json({text: result.text});
          }
        });
      }
    );
    req.on('error', next); // forwarded to error handler middleware in ErrorUtils.handle
    // Send the chat request to cohere
    req.write(JSON.stringify(chatBody));
    req.end();
  }

  private static createChatBody(body: Request['body']) {
    // Text messages are stored inside request body using the Deep Chat JSON format:
    // https://deepchat.dev/docs/connect
    return {
      query: body.messages[body.messages.length - 1].text,
      chat_history: body.messages.slice(0, body.messages.length - 1).map((message: {role: string; text: string}) => {
        return {user_name: message.role === 'ai' ? 'CHATBOT' : 'USER', text: message.text};
      }),
    };
  }

  public static async generateText(body: Request['body'], res: Response, next: NextFunction) {
    // Text messages are stored inside request body using the Deep Chat JSON format:
    // https://deepchat.dev/docs/connect
    const generationBody = {prompt: body.messages[0].text};
    const req = https.request(
      'https://api.cohere.ai/v1/generate',
      {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
          Authorization: 'Bearer ' + process.env.COHERE_API_KEY,
        },
      },
      (reqResp) => {
        let data = '';
        reqResp.on('error', next); // forwarded to error handler middleware in ErrorUtils.handle
        reqResp.on('data', (chunk) => {
          data += chunk;
        });
        reqResp.on('end', () => {
          const result = JSON.parse(data);
          if (result.message) {
            next(result); // forwarded to error handler middleware in ErrorUtils.handle
          } else {
            // Sends response back to Deep Chat using the Response format:
            // https://deepchat.dev/docs/connect/#Response
            res.json({text: result.generations?.[0].text});
          }
        });
      }
    );
    req.on('error', next); // forwarded to error handler middleware in ErrorUtils.handle
    // Send the chat request to cohere
    req.write(JSON.stringify(generationBody));
    req.end();
  }

  public static async summarizeText(body: Request['body'], res: Response, next: NextFunction) {
    // Text messages are stored inside request body using the Deep Chat JSON format:
    // https://deepchat.dev/docs/connect
    const summarizationBody = {text: body.messages[0].text};
    const req = https.request(
      'https://api.cohere.ai/v1/summarize',
      {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
          Authorization: 'Bearer ' + process.env.COHERE_API_KEY,
        },
      },
      (reqResp) => {
        let data = '';
        reqResp.on('error', next); // forwarded to error handler middleware in ErrorUtils.handle
        reqResp.on('data', (chunk) => {
          data += chunk;
        });
        reqResp.on('end', () => {
          const result = JSON.parse(data);
          if (result.message) {
            next(result); // forwarded to error handler middleware in ErrorUtils.handle
          } else {
            // Sends response back to Deep Chat using the Response format:
            // https://deepchat.dev/docs/connect/#Response
            res.json({text: result.summary});
          }
        });
      }
    );
    req.on('error', next); // forwarded to error handler middleware in ErrorUtils.handle
    // Send the chat request to cohere
    req.write(JSON.stringify(summarizationBody));
    req.end();
  }
}
